library/std/src/sys/pal/uefi/helpers.rs

```diff
@@ -221,3 +221,145 @@ pub(crate) fn runtime_services() -> Option<NonNull<r_efi::efi::RuntimeServices>>
+        for handle in handles {
+            if let Ok(protocol) = open_protocol::<r_efi::protocols::device_path_from_text::Protocol>(
+                handle,
+                r_efi::protocols::device_path_from_text::PROTOCOL_GUID,
+            ) {
+                LAST_VALID_HANDLE.store(handle.as_ptr(), Ordering::Release);
+                return inner(p, protocol);
+            }
+        }
+
+        io::Result::Err(const_io_error!(
+            io::ErrorKind::NotFound,
+            "DevicePathFromText Protocol not found"
+        ))
+    }
+}
+
+impl AsRef<r_efi::protocols::device_path::Protocol> for DevicePath {
+    fn as_ref(&self) -> &r_efi::protocols::device_path::Protocol {
+        unsafe { self.0.as_ref() }
+    }
+}
+
+impl AsMut<r_efi::protocols::device_path::Protocol> for DevicePath {
+    fn as_mut(&mut self) -> &mut r_efi::protocols::device_path::Protocol {
+        unsafe { self.0.as_mut() }
+    }
+}
+
+impl Drop for DevicePath {
+    fn drop(&mut self) {
+        if let Some(bt) = boot_services() {
+            let bt: NonNull<r_efi::efi::BootServices> = bt.cast();
+            unsafe {
+                ((*bt.as_ptr()).free_pool)(self.0.as_ptr() as *mut crate::ffi::c_void);
+            }
+        }
+    }
+}
+
+pub(crate) struct Protocol<T> {
+    guid: r_efi::efi::Guid,
+    handle: NonNull<crate::ffi::c_void>,
+    protocol: Box<T>,
+}
+
+impl<T> Protocol<T> {
+    const fn new(
+        guid: r_efi::efi::Guid,
+        handle: NonNull<crate::ffi::c_void>,
+        protocol: Box<T>,
+    ) -> Self {
+        Self { guid, handle, protocol }
+    }
+
+    pub(crate) fn create(protocol: T, mut guid: r_efi::efi::Guid) -> io::Result<Self> {
+        let boot_services: NonNull<r_efi::efi::BootServices> =
+            boot_services().ok_or(BOOT_SERVICES_UNAVAILABLE)?.cast();
+        let mut protocol = Box::new(protocol);
+        let mut handle: r_efi::efi::Handle = crate::ptr::null_mut();
+
+        let r = unsafe {
+            ((*boot_services.as_ptr()).install_protocol_interface)(
+                &mut handle,
+                &mut guid,
+                r_efi::efi::NATIVE_INTERFACE,
+                protocol.as_mut() as *mut T as *mut crate::ffi::c_void,
+            )
+        };
+
+        if r.is_error() {
+            return Err(crate::io::Error::from_raw_os_error(r.as_usize()));
+        };
+
+        let handle = NonNull::new(handle)
+            .ok_or(io::const_io_error!(io::ErrorKind::Uncategorized, "found null handle"))?;
+
+        Ok(Self::new(guid, handle, protocol))
+    }
+
+    pub(crate) fn handle(&self) -> NonNull<crate::ffi::c_void> {
+        self.handle
+    }
+}
+
+impl<T> Drop for Protocol<T> {
+    fn drop(&mut self) {
+        if let Some(bt) = boot_services() {
+            let bt: NonNull<r_efi::efi::BootServices> = bt.cast();
+            unsafe {
+                ((*bt.as_ptr()).uninstall_protocol_interface)(
```

Comment:
Will switch to `uninstall_multiple_protocol_interfaces` once https://github.com/r-efi/r-efi/pull/79 is merged.

Reply:
I dont think varargs are supported for eficall. I think it is only supported for "extern C". So the multi-call is not an option.

Reply:
Yes, but it seems `uninstall_multiple_protocol_interfaces` (and `install_multiple_protocol_interfaces`) are to be preferred over their single counterparts due to EFI 1.1 extensions that allow tracking. Thus `uninstall_multiple_protocol_interfaces` only fails if `One of the protocol interfaces was not previously installed on Handle. `, at least according to spec.

Reply:
What I am saying is that `uninstall_multiple_protocol_interfaces` uses a variable number of arguments, which is not supported by stable Rust for `eficall`. The prototype provided by `r-efi` is not usable in its current form and can easily lead to ABI violations.

You can look at EDK2 and see how `uninstall_multiple_protocol_interfaces` is implemented, and simply copy the logic from there.

Reply:
Also see: https://github.com/rust-lang/rust/issues/100189

Since `std` can use unstable features, you can just provide your own `fn` type and transmute the fn-pointer from the boot-services table.